platform/solr/solr-password/src/main/java/ddf/platform/solr/security/SolrPasswordUpdateImpl.java

```diff
@@ -0,0 +1,210 @@
+   */
+  private boolean isUsingDefaultPassword() {
+    return getPlaintextPasswordFromProperties().equals("autogenerated");
+  }
+
+  private @Nullable String getPlaintextPasswordFromProperties() {
+    String property = properties.getProperty(SOLR_PASSWORD_PROPERTY_NAME);
+    if (StringUtils.isBlank(property)) {
+      LOGGER.info(
+          "Solr password system property is missing or blank. System might not be able to communicate with Solr.");
+    }
+    return encryptionService.decryptValue(property);
+  }
+
+  /**
+   * Persist the encrypted password to the system properties file. This change does not take effect
+   * until the application is restarted.
+   */
+  private void setPasswordInPropertiesFile() {
+
+    String etcDir = properties.getProperty(KARAF_ETC);
+    String systemPropertyFilename = etcDir + File.separator + SYSTEM_PROPERTIES_FILE;
+    File systemPropertiesFile = new File(systemPropertyFilename);
+    AccessController.doPrivileged(
+        (PrivilegedAction<Void>)
+            () -> {
+              try {
+                Properties systemDotProperties = new Properties(systemPropertiesFile);
+                systemDotProperties.setProperty(
+                    SOLR_PASSWORD_PROPERTY_NAME, newPasswordWrappedEncrypted);
+                systemDotProperties.save();
+                LOGGER.info(
+                    "Updated encrypted Solr password in properties file {}.",
+                    systemPropertyFilename);
+              } catch (IOException e) {
+                LOGGER.error(
+                    "Exception while writing to {}. Solr password was changed, but new password was not saved.",
+                    systemPropertyFilename);
+              }
+              return null;
+            });
+  }
+
+  private void setPasswordInSolr() {
+    try (InputStream is = new ByteArrayInputStream(getSetUserJson().getBytes())) {
+      javax.ws.rs.core.Response response = solrAuthResource.sendRequest(is);
+      solrResponse = response.getStatusInfo();
+      if (isSolrPasswordChangeSuccessfull()) {
+        LOGGER.info("New password was set in Solr server.");
+      } else {
+        LOGGER.error(
+            "Solr password update failed with status code {}.", solrResponse.getStatusCode());
+      }
+    } catch (IOException e) {
+      LOGGER.info("Solr administration request failed.", e);
+    }
+  }
+
+  @VisibleForTesting
+  boolean isSolrPasswordChangeSuccessfull() {
+    return solrResponse != null && solrResponse.getFamily().equals(SUCCESSFUL);
+  }
+
+  /**
+   * Create a secure password. Use a UUID because it is long enough and random enough to withstand
+   * brute-force attacks.
+   */
+  @VisibleForTesting
+  void generatePassword() {
+    newPasswordPlainText = uuidGenerator.generateUuid();
+    newPasswordEncrypted = encryptionService.encrypt(newPasswordPlainText);
+    newPasswordWrappedEncrypted = String.format(ENC_TEMPLATE, newPasswordEncrypted);
```

Comment:
❗️ The knowledge of how a string is marked when encrypted shouldn't be a matter of public knowledge and therefore it should be here. The encryption service is the one that knows that and therefore, this formatting should be done there.

Reply:
I agree. I talked to @Lambeaux  about that. The interface is white-listed in the documentation, which I think that means we can't change it. However, we don't actually define what whitelisted means. Does it mean the interface can change, but not be deleted? Does it mean that existing methods cannot be removed, but new ones can be added? I'll ask around.

Reply:
That would be my take too.